### PR TITLE
IMDS Client

### PR DESCRIPTION
**Pull request by rcoh**
## Motivation and Context

IMDS credentials providers require an underlying IMDS client to function. A separate IMDS client is a commonly requested customer feature.

https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html

## Description
- Refactor TimeSource
- Refactor `Cache` so that it can be used with more than just credential
- Implement IMDS client and add lots of tests

Although it added a bit of code, I chose to use smithy_http::Operation and smithy_http::Client. This has the advantage of pulling in all the latest updates of logging / tracing / timeouts, etc. and routes all HTTP traffic through the same path.

## Testing
- [x] actually test this against real ECS
- [x] assorted simulation tests  

## Checklist
- [x] I have updated `CHANGELOG.md`
- [x] I have updated `aws/SDK_CHANGELOG.md` if applicable

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws/rust-runtime/aws-types/src/build_metadata.rs [hunk 1]
```diff
@@ -32,22 +31,21 @@ pub enum OsFamily {
 /// let os = get_cfg!(target_os: "linux", "android");
```
Comment (jdisanti) [inline aws/rust-runtime/aws-types/src/build_metadata.rs:31]: Macro usage instructions need an update

File: aws/rust-runtime/aws-types/src/build_metadata.rs [hunk 2]
```diff
@@ -56,21 +54,19 @@ impl From<&str> for OsFamily {
...
 }
 
-lazy_static! {
-    pub static ref BUILD_METADATA: BuildMetadata = BuildMetadata {
-        rust_version: RUST_VERSION,
-        core_pkg_version: env!("CARGO_PKG_VERSION"),
-        os_family: OsFamily::from_env()
-    };
-}
+pub const BUILD_METADATA: BuildMetadata = BuildMetadata {
```
Comment (jdisanti) [inline aws/rust-runtime/aws-types/src/build_metadata.rs:65]: Nice!

File: aws/rust-runtime/aws-types/src/os_shim_internal.rs [hunk 3]
```diff
@@ -208,11 +214,81 @@ mod env {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct TimeSource(time_source::Inner);
```
Comment (jdisanti) [inline aws/rust-runtime/aws-types/src/os_shim_internal.rs:218]: I really like this.

File: aws/rust-runtime/aws-types/src/os_shim_internal.rs [hunk 4]
```diff
@@ -208,11 +214,81 @@ mod env {
...
+
+impl Default for TimeSource {
+    fn default() -> Self {
+        TimeSource::real()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ManualTimeSource {
+    queries: Arc<Mutex<Vec<SystemTime>>>,
```
Comment (jdisanti) [inline aws/rust-runtime/aws-types/src/os_shim_internal.rs:245]: Shouldn't there be a way to access `queries` to confirm it in tests?

File: rust-runtime/smithy-client/src/bounds.rs [hunk 5]
```diff
@@ -123,7 +123,7 @@ pub trait SmithyRetryPolicy<O, T, E, Retry>:
     /// Forwarding type to `E` for bound inference.
     ///
     /// See module-level docs for details.
-    type E: Error + ProvideErrorKind;
+    type E: Error;
```
Comment (jdisanti) [inline rust-runtime/smithy-client/src/bounds.rs:126]: For my understanding, was this change needed so that the IMDS client could re-use the smithy-client retry code?

**Review by jdisanti — COMMENTED** [review]
Overall, I really like the changes here! I have a few minor questions and comments.

File: aws/rust-runtime/aws-config/src/imds/client.rs [hunk 6]
```diff
@@ -0,0 +1,893 @@
...
+
+/// An error retrieving metdata from IMDS
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum ImdsError {
+    /// An IMDSv2 Token could not be loaded
+    ///
+    /// Requests to IMDS must be accompanied by a token obtained via a `PUT` request. This is handled
+    /// transparently by the [`Client`].
+    FailedToLoadToken(SdkError<TokenError>),
```
[the remark this answers is not in the document]
Comment (rcoh) [reply aws/rust-runtime/aws-config/src/imds/client.rs:179]: since the error comes out of smithy-client it's broken down by phase and nested inside of `SdkError` (eg. error constructing the request etc.)

I could embed the SDK error inside the token error, but since we actually probably want the raw response for debugging, it seemed simplest to just leave it as an SdkError

File: aws/rust-runtime/aws-types/src/os_shim_internal.rs (hunk 4 above)
Comment (rcoh) [reply aws/rust-runtime/aws-types/src/os_shim_internal.rs:267]: yeah, will add. I hadn't added it since it wasn't needed yet.

File: aws/rust-runtime/aws-config/src/imds/client/token.rs [hunk 7]
```diff
@@ -133,7 +133,7 @@ impl TokenMiddleware {
             .call(operation)
             .await
             .map_err(ImdsError::FailedToLoadToken)?;
-        let expiry = response.expiry - TOKEN_REFRESH_BUFFER;
```
Comment (jdisanti) [inline aws/rust-runtime/aws-config/src/imds/client/token.rs]: Good catch! I forgot that the ExpiringCache was already considering the buffer time.

**Review by jdisanti — APPROVED** [review]
